docs/source/python_tutorials/fsa_algo.rst

```diff
@@ -0,0 +1,217 @@
+
+Its outputs are given below:
+
+.. figure:: images/before_invert.svg
+    :alt: before_invert
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/after_invert.svg
+    :alt: after_invert
+    :align: center
+    :figwidth: 600px
+
+
+When ``aux_labels`` is a ragged tensor, new arcs will be added after the invert.
+
+The following code applies :func:`k2.invert` to an FSA
+with ragged tensors as ``aux_labels``.
+
+.. literalinclude:: ./code/invert2.py
+  :caption: :func:`k2.invert` for FSAs with ragged tensors as ``aux_labels``
+  :language: python
+  :lines: 2-
+
+.. figure:: images/before_invert_aux.svg
+    :alt: before_invert_aux
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/after_invert_aux.svg
+    :alt: after_invert_aux
+    :align: center
+    :figwidth: 600px
+
+    Example of :func:`k2.invert` for FSAs with ragged tensors as ``aux_labels``.
+
+.. NOTE::
+
+  - :func:`k2.invert` supports CUDA as well as CPU.
+  - autograd is also supported.
+  - Its input can be either a single FSA or a FsaVec.
+
+
+arc_sort
+~~~~~~~~
+
+:func:`k2.intersect` and :func:`k2.compose` require two
+arc-sorted inputs. You can use :func:`k2.arc_sort` to convert
+an unsorted FSA/FsaVec into a sorted FSA/FsaVec.
+
+An FSA is sorted if for each state, its leaving arcs are sorted
+by ``labels`` in ascending order.
+
+.. CAUTION::
+
+  Arcs entering the final state have label -1. During sorting,
+  -1 is reinterpreted as an unsigned number. Internally, all
+  labels are reinterpreted as unsigned numbers during sorting.
+
+An example of :func:`k2.arc_sort` is given below.
+
+.. literalinclude:: ./code/arc_sort.py
+  :caption: :func:`k2.arc_sort` example
+  :language: python
+  :lines: 2-
+
+.. figure:: images/before_sort.svg
+    :alt: before_sort
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/after_sort.svg
+    :alt: after_sort
+    :align: center
+    :figwidth: 600px
+
+
+
+.. NOTE::
+
+  - :func:`k2.arc_sort` supports CUDA as well as CPU.
+  - autograd is also supported.
+  - Its input can be either a single FSA or a FsaVec.
+
+.. HINT::
+
+  Inside :func:`k2.arc_sort` it checks whether the input
+  FSA is sorted or not. If the input is already sorted, it
+  is returned directly. Otherwise, a new sorted FSA is returned.
+
+intersect
+~~~~~~~~~
+
+The following is an example of :func:`k2.intersect`.
+
+.. literalinclude:: ./code/intersect1.py
+  :caption: Example code for :func:`k2.intersect`
+  :language: python
+  :lines: 2-
+
+The outputs are shown below
+
+.. figure:: images/a_fsa_intersect.svg
+    :alt: a_fsa
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/b_fsa_intersect.svg
+    :alt: b_fsa
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/c_fsa_intersect.svg
+    :alt: c_fsa
+    :align: center
+    :figwidth: 600px
+
+:func:`k2.intersect` has an optional argument ``treat_epsilons_specially``.
+Its default value is `True`. If it is set to `False`, then the label `0`
```

Comment:
It may be better to clarify the beavior when `treat_epsilons_specially` is true or false, or add the link https://github.com/k2-fsa/k2/blob/9725fa2bd723acd13277249fa4719a851d9a36ea/k2/csrc/fsa_algo.h#L209 here?

Reply:
>  If it is set to False, then the label 0 is treated as a normal label.

In addition, an example is followed by the above statement. Do you have a better way to make it more understandable?

Reply:
I just feel our definition of `treat_epsilons_specially` has more meanings than users thought, user may be cannot figure out the meaning correctly from the examples. Anyway, adding the link to the documentation about `treat_epsilon_especially` Dan wrote in fsa_algo.h is fine with me.

Reply:
- when `treat_epsilons_specially` is True, 0 is treated as epsilons
- when `treat_epsilons_specially` is False, 0 is treated as normal labels

---


> I just feel our definition of treat_epsilons_specially has **more** meanings than users thought

What are the extra meanings besides the above ones?

Reply:
> when treat_epsilons_specially is True, 0 is treated as epsilons
    when treat_epsilons_specially is False, 0 is treated as normal labels

mm, we should clarify how the treat_epsilons_specially controls the beavior when matching labels in intersect. If you check the documentation @danpovey wrote in fsa_algo.h, you'll see that we also handle case where input changed by AddEpsilonSelfLoops. Anyway, I don't think users can get the meaining well above documentation.

Reply:
> You'll see that we also handle case where input changed by AddEpsilonSelfLoops.


I don't think there is code doing this kind of handling in `Intersect`. All it does is
- when treat_epsilons_specially is True, it also handles epsilon transitions
- when treat_epsilons_specially is False, epsilons are treated as normal labels. They are not special any more.

 
What @danpovey  documents is to explain in which case the behavior is correct if treat_epsilons_specially is False, I think.

---

> we should clarify how the treat_epsilons_specially controls the beavior when matching labels in intersect.

Can you explain why the following two statements confuse the users? I'd like to improve it.

- when treat_epsilons_specially is True, 0 is treated as epsilons
- when treat_epsilons_specially is False, 0 is treated as normal labels